### PR TITLE
*: add combination sql modes support(#4631)

### LINKED ISSUE #4631: `show variables` statement don't support "Combination SQL Modes"
Please answer these questions before submitting your issue. Thanks!

1. What did you do?

1). `show variables` statement don't support "Combination SQL Modes".
2). `show variables` statement don't remove duplicates.

my PR is on the way.

2. What did you expect to see?
```
mysql> set sql_mode='ANSI';
Query OK, 0 rows affected (0.00 sec)

mysql> show variables like 'SQL_MODE';
+---------------+-------------------------------------------------------------+
| Variable_name | Value                                                       |
+---------------+-------------------------------------------------------------+
| sql_mode      | REAL_AS_FLOAT,PIPES_AS_CONCAT,ANSI_QUOTES,IGNORE_SPACE,ANSI |
+---------------+-------------------------------------------------------------+
1 row in set (0.00 sec)
```

3. What did you see instead?
```
tidb>set sql_mode='ANSI';
Query OK, 0 rows affected (3 min 54.46 sec)

tidb>show variables like 'SQL_MODE';
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| sql_mode      | ANSI  |
+---------------+-------+
```

4. What version of TiDB are you using (`tidb-server -V`)?

```
Release Version: 0.9.0
Git Commit Hash: 376a7ddbddd0de91153ee36f6b6a434f22578363
Git Commit Branch: master
UTC Build Time:  2017-09-21 09:53:01
```

### PR DESCRIPTION
**Pull request by mccxj**
fixed issue #4631. add combination sql modes support.
see https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sql-mode-combo.

### CODE REVIEW

**Comment by shenli** [comment]
@mccxj Thanks for your PR!

File: mysql/const.go [hunk 1]
```diff
@@ -454,11 +454,23 @@ func FormatSQLModeStr(s string) string {
...
 			continue
 		}
-		nonEmptyParts = append(nonEmptyParts, parts[i])
+		if modeParts, ok := combinationSQLMode[part]; ok {
+			for _, modePart := range modeParts {
+				if _, exist := existParts[modePart]; !exist {
+					nonEmptyParts = append(nonEmptyParts, modePart)
+					existParts[modePart] = modePart
+				}
+			}
```
Comment (shenli) [inline mysql/const.go]: Could we continue here?

Comment (coocood) [reply mysql/const.go]: No, the combination mode itself is added too.

**Comment by coocood** [comment]
@mccxj 
Please fix CI.

**Comment by mccxj** [comment]
A little change. fix SQLMode is set wrongly when sqlmode is invalid.
```
tidb> insert into x2 select cast('999:44:33' as time);
ERROR 1292 (22007): Truncated incorrect time value: '999h44m33s'
tidb>set sql_mode=', ,';
ERROR 1105 (HY000): ERROR 1231 (42000): Variable 'sql_mode' can't be set to the value of ' '
tidb> insert into x2 select cast('999:44:33' as time);
Query OK, 1 row affected, 1 warning (0.00 sec)
```

**Comment by coocood** [comment]
@mccxj 
Please add a test for invalid SQL mode.

**Comment by mccxj** [comment]
@coocood testcase added. PTAL.

**Comment by coocood** [comment]
LGTM
@mccxj 
Thank you!

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@mccxj Thanks!

**Comment by mccxj** [comment]
@shenli Some test cases seem not very stable. eg. package "ddl".

**Comment by shenli** [comment]
@mccxj Yes, we will check it out. @zimulala

**Pull request merged**